File: src/code_review_agent/agents/lead_engineer.py
"""Lead Engineer synthesis agent.

Evaluates the aggregated outputs of the parallel review stage and produces
a final accept/reject decision for each finding.  The lead engineer does NOT
perform additional analysis beyond what the reviewers reported — its role is
to triage and prioritise, not to extend the review.
"""

import logging
from typing import ClassVar, cast

from strands import Agent
from strands.models.openai import OpenAIModel

from ..models.lead_engineer import (
    DecisionVerdict,
    FindingDecision,
    FindingDecisionOutput,
    LeadEngineerOutput,
    LeadEngineerReport,
)
from ..models.review import ReviewFinding, ReviewPerspective, ReviewReport
from .base_reviewer import ReviewerConfig

logger = logging.getLogger(__name__)

_SYSTEM_PROMPT = """\
You are a lead engineer responsible for triaging and prioritising code review
findings raised by a team of specialist reviewers.

Your sole task is to evaluate each finding submitted by the reviewers and
decide whether to accept it (the developer must address it) or reject it
(false positive, out of scope, or too low value to act on).

Decision criteria — consider all three axes:
1. Severity: How serious is the issue as reported by the reviewer?
2. Impact: What is the consequence of NOT fixing this issue?
3. Priority: How urgent is the fix relative to the PR goal?

Rules — you MUST follow every rule without exception:
- Base your decisions ONLY on the findings listed in the input.
- Do NOT introduce new issues, add inferred problems, or speculate beyond
  what the reviewers explicitly reported.
- Do NOT reference specific framework names or technology stacks in your
  reasoning unless a reviewer explicitly mentioned them.
- Every Finding in the input MUST receive a decision.
- Return the finding_index exactly as shown in the Finding # label.
- Assign final_priority; it may differ from the reviewer's original priority
  when the overall PR context justifies it.
- Provide a concise reason for each decision and an impact assessment.
"""


class LeadEngineerAgent:
    """Evaluates parallel reviewer outputs and produces final decisions.

    Consumes a :class:`~code_review_agent.models.review.ReviewReport` and
    returns a :class:`~code_review_agent.models.lead_engineer.LeadEngineerReport`
    with an accept/reject decision for every finding.

    The agent does NOT use GitHub MCP tools — its inputs are entirely derived
    from the reviewer outputs already collected.

    Args:
        config: Shared runtime configuration.  ``github_token`` is present for
            interface consistency but unused by this agent; ``model_id`` selects
            the LLM.
    """

    system_prompt: ClassVar[str] = _SYSTEM_PROMPT

    def __init__(self, config: ReviewerConfig) -> None:
        self._config = config

    def evaluate(self, report: ReviewReport) -> LeadEngineerReport:
        """Evaluate all reviewer findings and produce a final report.

        Args:
            report: Aggregated output from the parallel review stage.

        Returns:
            Final report with accept/reject decisions for every finding.
        """
        prompt, index_map = self._build_prompt_and_index(report)
        if self._config.llm_base_url:
            model = OpenAIModel(
                model_id=self._config.model_id,
                client_args={"base_url": self._config.llm_base_url},
            )
        else:
            model = OpenAIModel(model_id=self._config.model_id)
        agent = Agent(model=model, system_prompt=self.system_prompt, tools=[])
        output: LeadEngineerOutput = cast(
            LeadEngineerOutput,
            agent(
                prompt,
                structured_output_model=LeadEngineerOutput,
                limits={"turns": self._config.max_agent_turns},
            ).structured_output,
        )
        decisions = self._resolve_decisions(output.decisions, index_map)
        return LeadEngineerReport(
            overall_summary=output.overall_summary,
            decisions=decisions,
            reviewer_errors=report.errors,
        )

    def _build_prompt_and_index(
        self, report: ReviewReport
    ) -> tuple[str, dict[int, tuple[str, ReviewPerspective, ReviewFinding]]]:
        """Build the evaluation prompt and a finding-index map simultaneously.

        Each finding is assigned a 1-based index (``Finding #N``) so the LLM
        can reference it by number without reproducing the full finding object.
        The index map allows the agent code to look up the original finding
        after receiving the LLM's ``FindingDecisionOutput``.

        Args:
            report: Aggregated output from the parallel review stage.

        Returns:
            A tuple of (prompt string, index map).  The index map has the form
            ``{N: (reviewer_id, perspective, finding)}``.
        """
        lines: list[str] = [
            "Below are the findings from the parallel review stage.",
            "Evaluate each finding and produce a FindingDecision.",
            "",
        ]
        index_map: dict[int, tuple[str, ReviewPerspective, ReviewFinding]] = {}

        if not report.results:
            lines += [
                "No reviewer findings were submitted.",
                "",
                "Produce an overall_summary noting the absence of findings "
                "and an empty decisions list.",
            ]
            return "\n".join(lines), index_map

        n = 1
        for result in report.results:
            lines.append(
                f"=== Reviewer: {result.reviewer_id} "
                f"(perspective: {result.perspective.value}) ==="
            )
            lines.append(f"Reviewer summary: {result.output.summary}")
            lines.append("")

            if not result.output.findings:
                lines.append("  (no findings reported by this reviewer)")
                lines.append("")
                continue

            for finding in result.output.findings:
                index_map[n] = (result.reviewer_id, result.perspective, finding)
                lines.append(f"Finding #{n}")
                lines.append(f"  reviewer_id: {result.reviewer_id}")
                lines.append(f"  perspective: {result.perspective.value}")
                if finding.file_path:
                    lines.append(f"  file: {finding.file_path}")
                if finding.line:
                    lines.append(f"  line: {finding.line}")
                lines.append(f"  priority: {finding.priority.value}")
                lines.append(f"  comment: {finding.comment}")
                if finding.context:
                    lines.append(f"  context: {finding.context}")
                if finding.proposed_fix:
                    lines.append(f"  proposed_fix: {finding.proposed_fix}")
                lines.append("")
                n += 1

        lines.append(
            f"Produce exactly {len(index_map)} FindingDecision(s) — one per "
            "Finding listed above.  Do NOT add decisions for findings not listed."
        )
        return "\n".join(lines), index_map

    def _resolve_decisions(
        self,
        raw: list[FindingDecisionOutput],
        index_map: dict[int, tuple[str, ReviewPerspective, ReviewFinding]],
    ) -> list[FindingDecision]:
        """Resolve LLM output indexes to original findings.

        Normalises the LLM output to guarantee exactly one decision per finding
        in ``index_map``:

        1. Unknown indexes (not in ``index_map``) are logged and skipped.
        2. Duplicate indexes use only the first occurrence; later duplicates are
           logged and discarded.
        3. Findings with no LLM decision receive a deterministic default:
           ``REJECT`` with ``final_priority`` set to the original
           ``finding.priority``.

        Args:
            raw: LLM-generated decision outputs using finding indexes.
            index_map: Map from finding index to (reviewer_id, perspective,
                finding) built by :meth:`_build_prompt_and_index`.

        Returns:
            Resolved decisions with exactly one entry per finding in
            ``index_map``.
        """
        decisions: list[FindingDecision] = []
        seen: set[int] = set()

        for d in raw:
            entry = index_map.get(d.finding_index)
            if entry is None:
                logger.warning(
                    "LeadEngineerAgent: unknown finding_index %d — skipped",
                    d.finding_index,
                )
                continue
            if d.finding_index in seen:
                logger.warning(
                    "LeadEngineerAgent: duplicate finding_index %d — using first occurrence",
                    d.finding_index,
                )
                continue
            seen.add(d.finding_index)
            reviewer_id, perspective, finding = entry
            decisions.append(
                FindingDecision(
                    reviewer_id=reviewer_id,
                    perspective=perspective,
                    finding=finding,
                    verdict=d.verdict,
                    reason=d.reason,
                    impact=d.impact,
                    final_priority=d.final_priority,
                )
            )

        for idx, (reviewer_id, perspective, finding) in index_map.items():
            if idx not in seen:
                logger.warning(
                    "LeadEngineerAgent: finding_index %d has no LLM decision — defaulting to REJECT",
                    idx,
                )
                decisions.append(
                    FindingDecision(
                        reviewer_id=reviewer_id,
                        perspective=perspective,
                        finding=finding,
                        verdict=DecisionVerdict.REJECT,
                        reason="No decision provided by lead engineer.",
                        impact="Unknown — no evaluation provided.",
                        final_priority=finding.priority,
                    )
                )

        return decisions

File: packages/agent-core/src/agents/lead-engineer.ts
/**
 * Lead Engineer synthesis stage.
 *
 * Evaluates the aggregated parallel-review output and produces a final
 * accept/reject decision for every finding. Schemas and report-formatting
 * (`toMarkdown`/`toEvaluationFormat`) already live in `models/lead-engineer.ts`
 * (ported in an earlier slice); this module is the execution class only --
 * the direct counterpart of `LLMReviewAgent` for the synthesis stage.
 */

import { Agent } from "@strands-agents/sdk";
import {
  DecisionVerdict,
  type FindingDecision,
  type FindingDecisionOutput,
  FindingImpact,
  FindingPriority,
  type FindingSeverity,
  type LeadEngineerOutput,
  LeadEngineerOutputSchema,
  type LeadEngineerReport,
} from "../models/lead-engineer.js";
import type { ReviewFinding, ReviewPerspective, ReviewReport } from "../models/review.js";
import { ReviewPerspective as ReviewPerspectiveEnum } from "../models/review.js";
import type { ReviewerConfig } from "./base-reviewer.js";
import { StructuredOutputMissingError } from "./exceptions.js";
import { createModelProvider } from "./model-provider-factory.js";

const SYSTEM_PROMPT = `\
You are a lead engineer responsible for triaging and prioritising code review
findings raised by a team of specialist reviewers.

Your sole task is to evaluate each finding submitted by the reviewers and
decide whether to accept it (the developer must address it) or reject it
(false positive, out of scope, or too low value to act on).

Decision criteria -- consider all three axes:
1. Severity: How serious is the issue as reported by the reviewer?
2. Impact: What is the consequence of NOT fixing this issue?
3. Priority: How urgent is the fix relative to the PR goal?

Rules -- you MUST follow every rule without exception:
- Base your decisions ONLY on the findings listed in the input.
- Do NOT introduce new issues, add inferred problems, or speculate beyond
  what the reviewers explicitly reported.
- Do NOT reference specific framework names or technology stacks in your
  reasoning unless a reviewer explicitly mentioned them.
- Every Finding in the input MUST receive a decision.
- Return findingIndex as a plain integer matching the Finding # label --
  for "Finding #1" return 1, not "Finding #1" or "1" (a string). It must
  always be a JSON number.
- Assign severity, impactCategory, and finalPriority independently.
- severity must be critical, high, medium, or low.
- impactCategory must be security, correctness, performance, or maintainability.
- finalPriority must be high, medium, or low; it may differ from the reviewer's
  original priority when the overall PR context justifies it.
- Provide a concise reason for each decision and a prose impact assessment.`;

interface IndexEntry {
  reviewerId: string;
  perspective: ReviewPerspective;
  finding: ReviewFinding;
}

/**
 * Build the evaluation prompt and a finding-index map simultaneously.
 *
 * Each finding is assigned a 1-based index (`Finding #N`), numbered
 * consecutively across all reviewers in a single pass, so the LLM can
 * reference it by number without reproducing the full finding object.
 */
export function buildPromptAndIndex(report: ReviewReport): {
  prompt: string;
  indexMap: Map<number, IndexEntry>;
} {
  const lines: string[] = [
    "Below are the findings from the parallel review stage.",
    "Evaluate each finding and produce a FindingDecision.",
    "",
  ];
  const indexMap = new Map<number, IndexEntry>();

  if (report.results.length === 0) {
    lines.push(
      "No reviewer findings were submitted.",
      "",
      "Produce an overallSummary noting the absence of findings " + "and an empty decisions list.",
    );
    return { prompt: lines.join("\n"), indexMap };
  }

  let n = 1;
  for (const result of report.results) {
    lines.push(`=== Reviewer: ${result.reviewerId} (perspective: ${result.perspective}) ===`);
    lines.push(`Reviewer summary: ${result.output.summary}`);
    lines.push("");

    if (result.output.findings.length === 0) {
      lines.push("  (no findings reported by this reviewer)");
      lines.push("");
      continue;
    }

    for (const finding of result.output.findings) {
      indexMap.set(n, { reviewerId: result.reviewerId, perspective: result.perspective, finding });
      lines.push(`Finding #${n}`);
      lines.push(`  reviewerId: ${result.reviewerId}`);
      lines.push(`  perspective: ${result.perspective}`);
      if (finding.filePath) {
        lines.push(`  file: ${finding.filePath}`);
      }
      if (finding.line) {
        lines.push(`  line: ${finding.line}`);
      }
      lines.push(`  priority: ${finding.priority}`);
      lines.push(`  comment: ${finding.comment}`);
      if (finding.context) {
        lines.push(`  context: ${finding.context}`);
      }
      if (finding.proposedFix) {
        lines.push(`  proposedFix: ${finding.proposedFix}`);
      }
      lines.push("");
      n += 1;
    }
  }

  lines.push(
    `Produce exactly ${indexMap.size} FindingDecision(s) -- one per ` +
      "Finding listed above. Do NOT add decisions for findings not listed.",
  );
  return { prompt: lines.join("\n"), indexMap };
}

/**
 * Resolve LLM output indexes to original findings.
 *
 * Guarantees exactly one decision per finding in `indexMap`: unknown indexes
 * are logged and skipped, duplicate indexes keep only the first occurrence,
 * and a finding the LLM never covered gets a deterministic REJECT default
 * derived from the finding's own priority/perspective.
 */
export function resolveDecisions(
  raw: readonly FindingDecisionOutput[],
  indexMap: ReadonlyMap<number, IndexEntry>,
): FindingDecision[] {
  const decisions: FindingDecision[] = [];
  const seen = new Set<number>();

  for (const d of raw) {
    const entry = indexMap.get(d.findingIndex);
    if (entry === undefined) {
      console.warn(`LeadEngineerAgent: unknown findingIndex ${d.findingIndex} -- skipped`);
      continue;
    }
    if (seen.has(d.findingIndex)) {
      console.warn(
        `LeadEngineerAgent: duplicate findingIndex ${d.findingIndex} -- using first occurrence`,
      );
      continue;
    }
    seen.add(d.findingIndex);
    decisions.push({
      reviewerId: entry.reviewerId,
      perspective: entry.perspective,
      finding: entry.finding,
      verdict: d.verdict,
      reason: d.reason,
      impact: d.impact,
      severity: d.severity,
      impactCategory: d.impactCategory,
      finalPriority: d.finalPriority,
    });
  }

  for (const [idx, entry] of indexMap) {
    if (seen.has(idx)) {
      continue;
    }
    console.warn(
      `LeadEngineerAgent: findingIndex ${idx} has no LLM decision -- defaulting to REJECT`,
    );
    decisions.push({
      reviewerId: entry.reviewerId,
      perspective: entry.perspective,
      finding: entry.finding,
      verdict: DecisionVerdict.enum.REJECT,
      reason: "No decision provided by lead engineer.",
      impact: "Unknown -- no evaluation provided.",
      severity: entry.finding.priority as FindingSeverity,
      impactCategory:
        entry.perspective === ReviewPerspectiveEnum.enum.SECURITY
          ? FindingImpact.enum.SECURITY
          : FindingImpact.enum.CORRECTNESS,
      finalPriority:
        entry.finding.priority === "critical"
          ? FindingPriority.enum.HIGH
          : (entry.finding.priority as FindingPriority),
    });
  }

  return decisions;
}

/**
 * Evaluates parallel reviewer outputs and produces final decisions.
 *
 * Does NOT use GitHub MCP tools -- its inputs are entirely derived from the
 * reviewer outputs already collected. `config.githubToken` is present for
 * interface consistency but unused; `config.modelId` selects the LLM.
 */
export class LeadEngineerAgent {
  readonly systemPrompt = SYSTEM_PROMPT;

  constructor(private readonly config: ReviewerConfig) {}

  /**
   * Evaluate all reviewer findings and produce a final report.
   *
   * @throws StructuredOutputMissingError when the agent ends its turn
   *   without invoking the forced structured-output tool.
   */
  async evaluate(report: ReviewReport): Promise<LeadEngineerReport> {
    const { prompt, indexMap } = buildPromptAndIndex(report);
    const model = createModelProvider(
      this.config.providerType ?? "openai",
      this.config.modelId ?? "gpt-4o",
      {
        llmBaseUrl: this.config.llmBaseUrl,
        temperature: 0.3,
        maxTokens: this.config.maxTokens,
        frequencyPenalty: this.config.frequencyPenalty,
      },
    );
    const agent = new Agent({ model, systemPrompt: this.systemPrompt, tools: [] });
    const result = await agent.invoke(prompt, {
      structuredOutputSchema: LeadEngineerOutputSchema,
      limits: { turns: this.config.maxAgentTurns ?? 30 },
    });

    if (result.structuredOutput === undefined) {
      throw new StructuredOutputMissingError("LeadEngineerAgent", result.stopReason);
    }

    const output = result.structuredOutput as LeadEngineerOutput;
    const decisions = resolveDecisions(output.decisions, indexMap);
    return {
      overallSummary: output.overallSummary,
      decisions,
      reviewerErrors: report.errors,
    };
  }
}
